### PR TITLE
Made load balancer and public ip extended location aware.

### PR DESCRIPTION
**Pull request by MirzaSikander**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds extended location to the cloud config. 
If extended location is provided and a service is created, the corresponding slb and public ip are deployed with extended location.

**Special notes for your reviewer**:

**Release note**:
```
Adds extended location to the cloud config. 
If extended location is provided and a service is created, the corresponding slb and public ip are deployed with extended location.
```

### CODE REVIEW

File: pkg/provider/azure.go [hunk 1]
```diff
@@ -236,6 +240,11 @@ type Config struct {
 	Tags string `json:"tags,omitempty" yaml:"tags,omitempty"`
 }
 
+// HasExtendedLocation returns true if extendedlocation prop are specified.
+func (config *Config) HasExtendedLocation() bool {
+	return config.ExtendedLocationName != "" && config.ExtendedLocationType != ""
```
Comment (andyzhangx) [inline pkg/provider/azure.go:245]: `ExtendedLocationType` is redundant, if `ExtendedLocationName` is non empty, then use extended location.

File: pkg/provider/azure_loadbalancer.go [hunk 2]
```diff
@@ -823,6 +835,12 @@ func (az *Cloud) ensurePublicIPExists(service *v1.Service, pipName string, domai
 		}
 		pip.Name = to.StringPtr(pipName)
 		pip.Location = to.StringPtr(az.Location)
+		if az.HasExtendedLocation() {
+			pip.ExtendedLocation = &network.ExtendedLocation{
+				Name: &az.ExtendedLocationName,
+				Type: &az.ExtendedLocationType,
+			}
+		}
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:843]: If I bind a service to an existing pip, the pip would not have the extended location information on itself. Is this an expected behaviour?

File: pkg/provider/azure.go [hunk 3]
```diff
@@ -133,6 +133,10 @@ type Config struct {
 	ResourceGroup string `json:"resourceGroup,omitempty" yaml:"resourceGroup,omitempty"`
 	// The location of the resource group that the cluster is deployed in
 	Location string `json:"location,omitempty" yaml:"location,omitempty"`
+	// The name of site where the cluster will be deployed to that is more granular than the region sepecified by the "location" field.
+	ExtendedLocationName string `json:"extendedLocationName,omitempty" yaml:"extendedLocationName,omitempty"`
+	// The type of site that is being targeted.
+	ExtendedLocationType string `json:"extendedLocationType,omitempty" yaml:"extendedLocationType,omitempty"`
```
Comment (feiskyer) [inline pkg/provider/azure.go:139]: Could you add comments that these two new configs are only applied to LoadBalancer and PublicIP?

File: pkg/provider/azure.go (hunk 1 above)
Comment (feiskyer) [reply pkg/provider/azure.go:245]: seems both config are used for LB and PIP. If so, we should check both.

File: pkg/provider/azure_loadbalancer.go (hunk 2 above)
Comment (feiskyer) [reply pkg/provider/azure_loadbalancer.go:843]: Same qustion here, would it fail or still succeed in such a scenario?

File: pkg/provider/azure.go (hunk 3 above)
Comment (MirzaSikander) [reply pkg/provider/azure.go:139]: They will be applied to other things as well mainly persistent volumes

File: pkg/provider/azure_loadbalancer.go (hunk 2 above)
Comment (MirzaSikander) [reply pkg/provider/azure_loadbalancer.go:843]: Just a little context, when creating an AKS cluster, the user will choose the extended location, and the whole the cluster will be created in the edge zone. So all the initial resources will have extended location fields set. I am assuming that by existing PIP, you mean the PIP created by AKS? 

In any case, there are 2 creators, AKS RP or cloudprovider, in both cases extended Location will be set.

Comment (nilo19) [reply pkg/provider/azure_loadbalancer.go:843]: yeah I mean the unmanaged cluster created by aks-engine instead of aks

Comment (MirzaSikander) [reply pkg/provider/azure_loadbalancer.go:843]: We decided that AKS-engine will not support edge zones. That means the cloud config will never have the extended location properties set so we will never be in the situation

File: pkg/provider/azure.go (hunk 3 above)
Comment (feiskyer) [reply pkg/provider/azure.go:141]: thanks for confirming and updated comments

File: pkg/provider/azure_loadbalancer.go (hunk 2 above)
Comment (feiskyer) [reply pkg/provider/azure_loadbalancer.go:843]: Looks good. Could you squash the commits?

Comment (MirzaSikander) [reply pkg/provider/azure_loadbalancer.go:843]: Will do

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**